### PR TITLE
RFC about Tenant / Timeline guard objects

### PR DESCRIPTION
**Pull request by hlinnaka**

### CODE REVIEW

**Comment by hlinnaka** [comment]
Another way to think about this is that we need something like Arc, but with the extra capability to *wait* until there are no references to it, and to signal all current reference holders to request them to shut down. Some smart use of Arc/Weak and Drop might do the trick.

File: docs/rfcs/018-tenant-timeline-lifecycles.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+the code checks that the tenant is in Active state. If it's not, you get an
+error. You can change the state of the tenant to Stopping while there are
+ActiveTenantGuard objects still on it, to prevent new ActiveTenantGuards from
+being acquired, but the Tenant cannot be removed until all the guards are gone.
+
+TenantMaintenanceGuard: Like ActiveTenantGuard, but can be held even when the
+tenant is not in Active state. Used for operations like attach/detach. Perhaps
+allow only one such guard on a Tenant at a time.
+
+Similarly for Timelines. We don't currentl have a "state" on Timeline, but I think
```
Comment (koivunej) [inline docs/rfcs/018-tenant-timeline-lifecycles.md:77]:
```suggestion
Similarly for Timelines. We don't currently have a "state" on Timeline, but I think
```

**Review by koivunej — APPROVED** [review]
Leaning more on the type system sounds like a great idea, similar to i/o ownership which landed in 1.63 for file descriptors ([OwnedFd](https://doc.rust-lang.org/stable/std/os/unix/io/struct.OwnedFd.html)).

File: docs/rfcs/018-tenant-timeline-lifecycles.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+still some code operating on it.
+
+Let's introduce new Guard objects for Tenant and Timeline, and do all actions through
+the Guard object. Something like:
+
+TenantActiveGuard: Guard object over Arc<Tenant>. When you acquire the guard,
+the code checks that the tenant is in Active state. If it's not, you get an
+error. You can change the state of the tenant to Stopping while there are
+ActiveTenantGuard objects still on it, to prevent new ActiveTenantGuards from
+being acquired, but the Tenant cannot be removed until all the guards are gone.
```
Comment (SomeoneToIgnore) [inline docs/rfcs/018-tenant-timeline-lifecycles.md:71]: > but the Tenant cannot be removed until all the guards are gone

What if we invert?

Guard could hold `Weak<Tenant>` and is required to upgrade it to `Arc` (without exposing the `Arc` itself to the API callers) as one of the extra checks.

**Review by SomeoneToIgnore — COMMENTED** [review]
I like the idea, have considered once similar things following the `Weak` trail: so far the biggest downside with the whole idea is a hard requirement to wrap `Timeline` API around such objects (`Guard`) since otherwise `Arc` could leak and can be easily cloned.

**Pull request closed without merging**